### DISCUSSION TITLE
Reading and interpolating WOA monthly climatology files for Grid_Stat

### DISCUSSION THREAD

**Discussion by GwenChen-NOAA** [dtcenter/METplus#1604] [Configuration] [answered]
I used the following GRID_STAT_CLIMO_MEAN setup in my METplus 4.0.0 configuration file to read in WOA monthly climatology files, regrid them to match OBS grids, and interpolate them to the VALID date:

GRID_STAT_CLIMO_MEAN_FIELD = {name = "t_an"; level = "(0,0,*,*)";}     --- please note star,star may not show up in the text
GRID_STAT_CLIMO_MEAN_REGRID_METHOD = BILIN
GRID_STAT_CLIMO_MEAN_REGRID_WIDTH = 2
GRID_STAT_CLIMO_MEAN_REGRID_VLD_THRESH = 0.5
GRID_STAT_CLIMO_MEAN_TIME_INTERP_METHOD = DW_MEAN
GRID_STAT_CLIMO_MEAN_DAY_INTERVAL = 31
GRID_STAT_CLIMO_MEAN_HOUR_INTERVAL = 6
GRID_STAT_CLIMO_MEAN_INPUT_DIR = {ENV[CLIMO_PATH]}/WOA18
GRID_STAT_CLIMO_MEAN_INPUT_TEMPLATE = woa18_decav_t{valid?fmt=%m}_04.nc

When I ran METplus, it pointed to the correct climatology file, but it can't read the WOA file with the following error message in the log file:

ERROR  : DictionaryEntry::dict_value() -> bad type

The CLIMO_MEAN dictionary is outputted in the log file as:

metplus INFO: METPLUS_CLIMO_MEAN_DICT=climo_mean = {file_name = ["/lfs/h2/emc/vpppg/noscrub/Lichuan.Chen/Global/climo/WOA18/woa18_decav_t05_04.nc"];field = ["{name=\"t_an\"; level=\"(0,0,*,*)\";}"];regrid = {method = BILIN;width = 2;vld_thresh = 0.5;}time_interp_method = DW_MEAN;day_interval = 31;hour_interval = 6;}

--- I just noticed that the field didn't show up the same as I posted. There are backslashes before the " in the log text, so it's field = ["{name= backslash"t_an backslash"; level=backslash"(0,0,star,star)backslash";}"];

In which, the syntax for the field looks a bit strange to me.  Is this correct?  My FCST and OBS are both netCDF files with the following fields:

metplus INFO: METPLUS_FCST_FIELD=field = [{ name="sst"; level="(0,*,*)"; }];     --- it should be "(0,star,star)"
metplus INFO: METPLUS_OBS_FIELD=field = [{ name="analysed_sst"; level="(0,*,*)"; convert(x) = K_to_C(x); }];

WOA SST data are 4-dimensional: (time, depth, lat, lon), different from FCST and OBS, which are 3-dimensional: (time, lat, lon).  Is it a problem?  I can run Grid_Stat and generate CNT stat files using FCST and OBS without CLIMO_MEAN.  I also can read WOA climatology files using plot_data_plane with 'name = "t_an"; level = "(0,0,*,*)";' to create SST maps.  Is there anything I missed in setting up CLIMO_MEAN?  Thanks for your help!

**Comment by georgemccabe** [comment]
Hi @GwenChen-NOAA,

You have configured METplus correctly to read the climo mean field. However, enhancements made for METplus 4.1.0 fix the incorrect formatting that you are seeing with the quotation marks in the climo_mean field value. If you are able to upgrade to using version 4.1.0, I recommend making that switch to remedy this issue. If you must continue to use 4.0.0, then you should be able to get around this issue by using the GRID_STAT_MET_CONFIG_OVERRIDES variable instead of GRID_STAT_CLIMO_MEAN_FIELD. You would set something like this:

`GRID_STAT_MET_CONFIG_OVERRIDES = climo_mean = { field = [{name = "t_an"; level = "(0,0,*,*)";}];}`

This will add this text to the end of the MET configuration file and override the value that way.

On a related note, there is an open GitHub issue #1599 that is slated to be completed in the next development cycle (METplus 5.0.0-beta1 tentatively scheduled for 6/22) that will allow you set the field information this way:

```
GRID_STAT_CLIMO_MEAN_VAR1_NAME = t_an
GRID_STAT_CLIMO_MEAN_VAR1_LEVELS = "(0,0,*,*)"
```

The METplus wrapper logic will properly format the values into the correct format that is expected by the MET config file.

Please let me know if you have any other issues or questions. Thanks.

**Reply by GwenChen-NOAA** [reply]
George,

Thanks for the quick response and suggestions!  The GRID_STAT_MET_CONFIG_OVERRIDES works and Grid_Stat is reading the WOA file.  However, the time interpolation doesn't work and no anomalies were calculated.  I got the following error message in the log file:

DEBUG 4: Data plane information:
DEBUG 4:       plane min: -2.08301
DEBUG 4:       plane max: 31.2889
DEBUG 4:      valid time: 19700101_000000
DEBUG 4:       lead time: 000000
DEBUG 4:       init time: 19700101_000000
DEBUG 4:      accum time: 000000
DEBUG 4:
DEBUG 3: Skipping the 19700101_000000 "t_an(0,0,*,*)" climatology field since the time offset (10627200 seconds) >= the "day_interval" entry (2678400 seconds) from file: /lfs/h2/emc/vpppg/noscrub/Lichuan.Chen/Global/climo/WOA18/woa18_decav_t05_04.nc
DEBUG 3: Found 0 climatology fields.
DEBUG 3: Found 0 climatology mean and 0 climatology standard deviation field(s) for forecast sst(0,*,*).

My valid time for FCST is 20220504_000000.  Any suggestions how to fix it?  Thanks!

**Reply by georgemccabe** [reply]
Is the file you are trying to read a CF-compliant NetCDF file? It looks like the time information in the file is not being read by MET properly. Could you send me the file? There are instructions to upload the file to our FTP server here: #954

**Reply by GwenChen-NOAA** [reply]
George,

I have uploaded my data to the ftp server under the chen_data directory.  WOA climatology files are CF-compliant NetCDF files, but the time stamp in the files may not be meaningful because they are monthly mean averaged over a period of time.  The represented averaging period is denoted in the filename.  For example, the woa18_decav_t05_04.nc file is for May averaged over 1981 – 2010 for quarter-degree grid resolution.  I also uploaded the WOA documentation in the chen_data directory.  The file naming convention is on Page 12.  Thanks!

**Comment by georgemccabe** [comment, marked as answer]
Hi @GwenChen-NOAA, thank you for sending your data and the documentation. It looks like the units for the time variable is "months since 1955-01-01 00:00:00" and while that is technically a valid time unit for CF-compliance, the MET tools do not support it. We often recommend using the MET plot_data_plane tool to check if MET can read new input files properly. Running this command:

`plot_data_plane woa18_decav_t05_04.nc out.nc 'name="t_an"; level="(0,0,*,*)";'`

outputs this warning:

> WARNING: parse_cf_time_string() -> Unsupported time step in the CF convention time unit "months since 1955-01-01 00:00:00"

You can also run plot_data_plane with the "-v 4" argument to raise the logging verbosity. The grid and time information that is parsed from the file is output at the DEBUG 4 level.

Fortunately you can override the attributes that are read parsed by MET using the set_attr_* variables. In this case, you would want to set the valid time using set_attr_valid. You can use the filename template syntax to get the current valid time you are processing:

`GRID_STAT_MET_CONFIG_OVERRIDES = climo_mean = { field = [{name = "t_an"; level = "(0,0,*,*)"; set_attr_valid="{valid?fmt=%Y%m%d_%H%M%S}";}];}`

Search for "set_attr" on [this page](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=set_attr#fcst) to see all of the available attributes that you can override.

Let me know if that works for you.

**Reply by GwenChen-NOAA** [reply]
Thank you very much, George!  The GRID_STAT_MET_CONFIG_OVERRIDES with set_attr_valid works, and I got anomaly correlation outputted in the stat file.  If I set the valid time for the monthly mean climatology as the FCST verification time, will METplus perform any time interpolation?  We usually assume the monthly mean is centered on the 15th and then interpolate it to the valid date.  I tried to set the valid time to %Y%m15_%H%M%S (e.g., 20220515_000000), and I got identical results as I set it to %Y%m%d_%H%M%S (e.g., 20220504_000000.  Below is the outputs from my log file:

DEBUG 4: VarInfoFactory::new_var_info() -> created new VarInfo object of type "FileType_NcCF".
DEBUG 3: Parsed set_attr_valid = "20220515_000000"
DEBUG 3: Resetting valid time from 19700101_000000 to 20220515_000000.
DEBUG 4:
DEBUG 4: Data plane information:
DEBUG 4:       plane min: -2.1
DEBUG 4:       plane max: 31.2889
DEBUG 4:      valid time: 20220515_000000
DEBUG 4:       lead time: 000000
DEBUG 4:       init time: 19700101_000000
DEBUG 4:      accum time: 000000
DEBUG 4:
DEBUG 2: Regridding the 20220515_000000 "t_an(0,0,*,*)" climatology field to the verification grid.
DEBUG 3: For forecast valid at 20220504_000000, found 1 climatology field(s) with valid time(s): 20220515_000000
DEBUG 3: Found 1 climatology fields.
DEBUG 3: Found 1 climatology mean and 0 climatology standard deviation field(s) for forecast sst(0,*,*).
DEBUG 2: Processing masking regions.

I saw it found the climatology mean, but I didn't see any information about time interpolation.  Does the DW_MEAN work?  Or it's grabbing the nearest value?  Thanks!

**Reply by JohnHalleyGotway** [reply]
@GwenChen-NOAA unfortunately NO, MET would not be doing any time interpolation when using the ```set_attr_valid``` option. @georgemccabe and I discussed some of the details offline.

Let me state the issues as I understand them.

1. As George notes, MET doesn't support the "months since" time convention. The use of months is discouraged since they vary. According to this section of the [CF-convention documentation](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.9/cf-conventions.html#time-coordinate), 1 year = 365.242198781 days and 1 month = 1 year / 12. So 1 month = 30.436 days.

Looking at one of the sample files you sent, by that logic, MET should interpret 377.5 months since 1955-01-01 00:00:00 as Jan 1, 1955 + 11489.91 days = 1986-06-16 21:51:36.

Adding that logic based on the CF-conventions spec would be very easy. However, I strongly suspect this is the logic you would actually expect:
time = 377.5 = 377 months + 15 days = 31 years + 5 months + 15 days
By that logic, we'd get 1986-06-15 00:00:00.

Those 2 methods results in a difference of almost 2 days. Am I right in assuming you'd like the MET software to use the second date of 1986-06-15 00:00:00 instead of the first one?

2. Since MET isn't reading the timestamps from these files, then they all default to a valid time of 1970-01-01. Using "set_attr_valid" overrides that timestamp. However, that's overridden for ALL climo files it reads. If you're reading one, and only one, climo mean file, that patch could work. But if you want to read data from multiple files (which is necessary for any time interpolation) it won't work, because MET will set the valid time to the set_attr_valid value for all the climo files it reads.

So what do we do?

Options include...
- Create a complex python-embedding solution to handle this situation. I don't think that's worth it and will just make the situation more confusing.
- Write a script to modify your existing WOA climo data to redefine the time variable to use "days since..." instead of "months since...". This would enable you to get up and running with MET version 10.1.0.
- Use the existing WOA data but enhance MET with one of the methods listed above to handle the "months since..." convention. That would require a new MET version before you'd be able to make progress.

How would you like to proceed?

**Reply by GwenChen-NOAA** [reply]
John,

Thank you for the detailed explanation.  I think for the long term, it would be the best that MET/METplus can interpolate the existing WOA data as NCEI update this data set every 4-5 years, so it's easier to replace it in an operational system.  However, we plan to have an alpha version of EVS by the end of this year, so I may not have time to wait for a new MET version to make progress.  I can try the 2nd option first.  In your example for June, we'd like the MET software to use the second date of 1986-06-15 00:00:00.  If I'm able to redefine the time variable to use "days since...", what date should I set for the June climo file?  Will 1986-06-15 00:00:00 work?  Or a closer date (e.g., 2018-06-15 00:00:00) is more likely to be found by MET?  Thanks!

**Reply by JohnHalleyGotway** [reply]
@GwenChen-NOAA the year really doesn't matter. Usually climo datasets are defined over some long time period, perhaps 30 years or 40 years? And data producers will label the resulting timestamp using the first year or last year of that time interval.

The sample file you sent includes these global attributes:
```
		:time_coverage_start = "1955-06-01" ;
		:time_coverage_end = "2017-06-30" ;
```
It looks like 1986 is midway between those beginning and ending dates. So I assume they chose the mid-point. To minimize the changes, you could just use the exact same timestamp of 1986-06-15 00:00:00.

But you'd define that as being 11337 "days since 1986-06-15 00:00:00". FYI, I cheated and used [this website](https://www.timeanddate.com/date/durationresult.html?m1=6&d1=1&y1=1955&m2=6&d2=15&y2=1986) to compute that. But you should definitely be able to compute it using the date command in unix or datetime in python.

MET realizes the year value of the climo data will differ from the forecast valid time. Its looking for matching months, days, and hours.

**Reply by JohnHalleyGotway** [reply]
FYI, I wrote up this MET development issue to capture this feature request:
https://github.com/dtcenter/MET/issues/2155

**Reply by GwenChen-NOAA** [reply]
Thank you, John!  It's good to know that the year would not affect the DW_MEAN calculation.  I will let you know if I can work out the 2nd option.  Thanks!

**Reply by GwenChen-NOAA** [reply]
@JohnHalleyGotway, I used the below CDO command to change the time unit to days in the WOA climatology files:

cdo settunits,days -settaxis,1986-05-15,00:00,1month woa18_decav_t05_04.nc woa18_decav_t05_04.newtime.nc

Then rerun Grid_Stat.  MET is reading the timestamp correctly, but I still got identical results as before.  No DW_MEAN time interpolation was carried out.  The outputs in the log file are below.

DEBUG 4: Met2dDataFileFactory::new_met_2d_data_file() -> created new Met2dDataFile object of type "FileType_NcCF".
DEBUG 4: NcCfFile::open() -> parsing units for the time variable "days since 1986-5-15 00:00:00"
DEBUG 4: parse_cf_time_string() -> parsed NetCDF CF convention time unit string "days since 1986-5-15 00:00:00"
DEBUG 4:                 as a reference time of 19860515_000000 and 86400 second(s) per time step.
DEBUG 4: NcCfFile::open() -> could not extract init time from the "forecast_reference_time" variable.
DEBUG 4: NcCfFile::open() -> could not extract init time from file name.
DEBUG 4: NcCfFile::get_grid_from_dimensions() -> The coordinate variable "bnds" does not exist.
DEBUG 4: VarInfoFactory::new_var_info() -> created new VarInfo object of type "FileType_NcCF".
DEBUG 4: NcCfFile::getData() -> setting the unset init time to the valid time of 19860515_000000.
DEBUG 4:
DEBUG 4: Data plane information:
DEBUG 4:       plane min: -2.1
DEBUG 4:       plane max: 31.2889
DEBUG 4:      valid time: 19860515_000000
DEBUG 4:       lead time: 000000
DEBUG 4:       init time: 19860515_000000
DEBUG 4:      accum time: 000000
DEBUG 4:
DEBUG 2: Regridding the 19860515_000000 "t_an(0,0,*,*)" climatology field to the verification grid.
DEBUG 3: For forecast valid at 20220509_000000, found 1 climatology field(s) with valid time(s): 20220515_000000
DEBUG 3: Found 1 climatology fields.
DEBUG 3: Found 1 climatology mean and 0 climatology standard deviation field(s) for forecast sst(0,*,*).
DEBUG 2: Processing masking regions.

I also uploaded the woa18_decav_t05_04.newtime.nc file, my Grid_Stat configuration file, and METplus log file in the ftp site under chen_data.  I noticed in the log file, the climo_mean is outputted as: 

metplus INFO: METPLUS_CLIMO_MEAN_DICT=climo_mean = {file_name = ["/lfs/h2/emc/vpppg/noscrub/Lichuan.Chen/Global/climo/WOA18/woa18_decav_t05_04.newtime.nc"];regrid = {method = BILIN;width = 2;vld_thresh = 0.5;shape = SQUARE;}time_interp_method = DW_MEAN;day_interval = 31;hour_interval = 6;}

There is no ; before time_interp_method = DW_MEAN.  Does it hinder MET setting the time interpolation method?  If the 2nd option doesn't work, what could I do?  Please advise.  Thanks!

**Reply by JohnHalleyGotway** [reply]
@GwenChen-NOAA thanks for sending along the log files. I understand the issue that Grid-Stat is not interpolating the climo data to the valid time of the forecast. I'm hopeful that we can make this work by modifying the configuration files.

I want to point out this line in the METplus log file:
```
05/13 16:08:57.134 metplus INFO: METPLUS_CLIMO_MEAN_DICT=climo_mean = {file_name = ["/lfs/h2/emc/vpppg/noscrub/Lichuan.Chen/Global/climo/WOA18/woa18_decav_t05_04.newtime.nc"];regrid = {method = BILIN;width = 2;vld_thresh = 0.5;shape = SQUARE;}time_interp_method = DW_MEAN;day_interval = 31;hour_interval = 6;}
```
This tells us exactly how the "climo_mean" dictionary is configured for use by Grid-Stat. Note that we are telling Grid-Stat to read climo data from one, AND ONLY ONE, WOA file. Grid-Stat obviously needs to read the climo data from before and after the valid time of the forecast to be able to interpolate to it.

We need to configure Grid-Stat to do so. I think it'd be simplest to pass in 3 files each time, for the current month, the previous month, and the next month. That way we don't have to worry about being before/after the 15th of the month. So here's the goal for April, May, June:
```
file_name = ["/lfs/h2/emc/vpppg/noscrub/Lichuan.Chen/Global/climo/WOA18/woa18_decav_t04_04.newtime.nc", "/lfs/h2/emc/vpppg/noscrub/Lichuan.Chen/Global/climo/WOA18/woa18_decav_t05_04.newtime.nc", "/lfs/h2/emc/vpppg/noscrub/Lichuan.Chen/Global/climo/WOA18/woa18_decav_t06_04.newtime.nc"];
```

I have two questions.

1. How do we configure METplus to have the desired effect? Hopefully we can do so using the "shift" option, like this:
```
GRID_STAT_CLIMO_MEAN_INPUT_TEMPLATE = woa18_decav_t{valid?fmt=%m?shift=-30d}_04.newtime.nc, woa18_decav_t{valid?fmt=%m}_04.newtime.nc, woa18_decav_t{valid?fmt=%m?shift=30d}_04.newtime.nc
```
I'm subtracting/adding 30 days to get the previous/next months.

2. When users request ```time_interp_mthd = DW_MEAN;``` (or UW_MEAN) but provide only 1 climo file, perhaps that should result in a warning or error message. What do you think about that?

Please let me know if updating the config file make the time interpolation work.

**Reply by JohnHalleyGotway** [reply]
@GwenChen-NOAA I also wanted to point out this warning message I see in the output you sent:
```
WARNING: process_scores() -> Forecast and observation valid times do not match 20220509_000000 != 20220509_120000 for sst(0,*,*) versus analysed_sst(0,*,*).
```
Are you aware of the discrepancy in the timestamps of the fcst and obs data?

**Reply by GwenChen-NOAA** [reply]
@JohnHalleyGotway , Thanks for the suggestion. I changed the setting to GRID_STAT_CLIMO_MEAN_INPUT_TEMPLATE = woa18_decav_t{valid?fmt=%m?shift=-30D}_04.newtime.nc, woa18_decav_t{valid?fmt=%m}_04.newtime.nc, woa18_decav_t{valid?fmt=%m?shift=30D}_04.newtime.nc in my configuration file, but I got fatal error when running METplus. I think METplus cannot recognize 3 files in the template. I uploaded my log file (metplus.log.20220516215705) in the ftp site under chen_data. I also put a RTOFS forecast file, a GHRSST data file, and 3 WOA climatology files in the same directory.  Is it possible to specify 3 files in GRID_STAT_CLIMO_MEAN_FILE_NAME? If so, I can set up the valid month and the months before and after as environment variables in my scripts and then pass them into the Grid_Stat configuration file to see if it works.

For the discrepancy in the timestamps of the fcst and obs data, I did see the warning. It is a known issue in the current RTOFS Satellite Verification system. Ideally, we would like the hours matched for both fcst and obs data, but it is not an easy task computationally. Fortunately, ocean variables evolve slowly, so it is safe to assume that the differences between 00Z and 12Z are negligible. In the current system, we simply shift the time of the obs data to 00Z to match the valid time. METplus seems to be tolerant of the discrepancy and still outputs numbers in the CNT stat file despite the warning, so I leave it as is for now. Thanks!

**Reply by JohnHalleyGotway** [reply]
I suspect I gave you bad advice. Please try changing the capital 'D' to a lowercase one 'd'. I searched the METplus repo and found an example of shifting by a number of days, like this ```{valid?fmt=%Y%m%d_%H?shift=1d}``` with a lowercase 'd'.

So please try using this ```GRID_STAT_CLIMO_MEAN_INPUT_TEMPLATE = woa18_decav_t{valid?fmt=%m?shift=-30d}_04.newtime.nc, woa18_decav_t{valid?fmt=%m}_04.newtime.nc, woa18_decav_t{valid?fmt=%m?shift=30d}_04.newtime.nc```

If that doesn't solve it, I'll ask @georgemccabe to weigh in.

**Reply by GwenChen-NOAA** [reply]
@JohnHalleyGotway ,  Good news! This one works! METplus is able to grab the correct monthly files and interpolate to the valid date. The log file has the following outputs:

DEBUG 2: Regridding the 19860615_000000 "t_an(0,0,*,*)" climatology field to the verification grid.
DEBUG 3: For forecast valid at 20220517_000000, found 2 climatology field(s) with valid time(s): 20220515_000000, 20220615_000000
DEBUG 3: Interpolating climatology fields at 20220515_000000 and 20220615_000000 to 20220517_000000 using the DW_MEAN interpolation method.
DEBUG 3: Found 1 climatology fields.
DEBUG 3: Found 1 climatology mean and 0 climatology standard deviation field(s) for forecast sst(0,*,*).
DEBUG 2: Processing masking regions.

For other people's reference, below is my climo_mean setup in the Grid_Stat configuration file for METplus 4.0.0:

GRID_STAT_MET_CONFIG_OVERRIDES = climo_mean = {field = [{name="t_an"; level="(0,0,*,*)";}];}
GRID_STAT_CLIMO_MEAN_REGRID_METHOD = BILIN
GRID_STAT_CLIMO_MEAN_REGRID_WIDTH = 2
GRID_STAT_CLIMO_MEAN_REGRID_VLD_THRESH = 0.5
GRID_STAT_CLIMO_MEAN_REGRID_SHAPE = SQUARE
GRID_STAT_CLIMO_MEAN_TIME_INTERP_METHOD = DW_MEAN
GRID_STAT_CLIMO_MEAN_DAY_INTERVAL = 31
GRID_STAT_CLIMO_MEAN_HOUR_INTERVAL = 6
GRID_STAT_CLIMO_MEAN_INPUT_DIR = {ENV[CLIMO_PATH]}/WOA18_newtime
GRID_STAT_CLIMO_MEAN_INPUT_TEMPLATE = woa18_decav_t{valid?fmt=%m?shift=-31d}_04.newtime.nc, woa18_decav_t{valid?fmt=%m}_04.newtime.nc, woa18_decav_t{valid?fmt=%m?shift=28d}_04.newtime.nc

Please note that I changed the shift days in the template for the month before from 30 days to 31days, so for valid date on the 31th (e.g., 20220531) it will grab the previous month (e.g., April), and for the month after to 28 days, so for valid date on Jan 31 it will grab February. This setup requires adjustment of the time unit from months to days in the WOA18 climatology files.

Thank you so much for all your help! I'm very happy we work this out. Thanks!

**Reply by JohnHalleyGotway** [reply]
That's great news! Thanks for confirming!

Can you please select one of the responses in this thread as a good answer and click the "Mark As Answer" button? Then I'll lock discussion to prevent future posts.

Thanks
